mithril-stm/src/stm.rs

```diff
@@ -879,6 +942,104 @@ impl<D: Clone + Digest + FixedOutput + Send + Sync> StmAggrSig<D> {
     }
 }
 
+impl<D: Digest + FixedOutput> FullNodeVerifier<D> {
+    /// Verify all signatures whether they all won the lottery, if the indices are unique and the quorum is achieved.
+    pub fn pre_verify(
+        total_stake: &Stake,
+        signatures: &[StmSig],
+        parameters: &StmParameters,
+        msg: &[u8],
+        signed_parties: Vec<RegParty>,
+    ) -> Result<(), StmAggregateSignatureError<D>> {
+        let mut nr_indices = 0;
+        let mut unique_indices = HashSet::new();
+
+        for (i, sig) in signatures.iter().enumerate() {
+            sig.check_indices(parameters, &signed_parties[i].1, msg, total_stake)?;
+            for &index in &sig.indexes {
+                unique_indices.insert(index);
+                nr_indices += 1;
+            }
+        }
```

Comment:
Now we are verifying signature wrt to keys from two different data structures (two independent slices). This is error prone, and a user could provide a list that is not ordered. Previously, the aggregate signature contained a vector that tupled the signature with it's corresponding key. Why are we no longer using that?